### ISSUE TITLE
editing functions causes redefinitions

### ISSUE THREAD

**Issue by fdncred** [sigmaSd/IRust#67]
I type this is in and run it and notice i've made a mistake.

```rust
In: fn factorial(n: size) -> usize {
..:     match n {
..:         1 => 1,
..:         n => n * factorial(n - 1),
..:     }
..: }

In: :type factorial
fn item `fn([type error]) -> usize {factorial}`
In: factorial(12)

error[E0412]: cannot find type `size` in this scope
 --> src\main.rs:2:17
  |
2 | fn factorial(n: size) -> usize {
  |                 ^^^^ help: a builtin type with a similar name exists: `isize`
```
So, I up-arrow and fix the mistake and run it again and now there are multiple definitions.

```rust
In: fn factorial(n: usize) -> usize {
..:     match n {
..:         1 => 1,
..:         n => n * factorial(n - 1),
..:     }
..: }

In: factorial(12)

error[E0428]: the name `factorial` is defined multiple times
 --> src\main.rs:8:1
  |
2 | fn factorial(n: size) -> usize {
  | ------------------------------ previous definition of the value `factorial` here
...
8 | fn factorial(n: usize) -> usize {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `factorial` redefined here
  |
  = note: `factorial` must be defined only once in the value namespace of this block

error[E0412]: cannot find type `size` in this scope
 --> src\main.rs:2:17
  |
2 | fn factorial(n: size) -> usize {
  |                 ^^^^ help: a builtin type with a similar name exists: `isize`
```

What I meant for it to do was overwrite the first one with the second one. Not create two.

**Comment by sigmaSd** [comment]
This is because of https://github.com/sigmaSd/IRust/issues/62
Every statement used to require ';' at the end, but now there are some exception that gets immediately parsed into the repl (like function definition see https://github.com/sigmaSd/IRust/blob/master/src/irust/parser.rs#L303)
I found overall the convenience of not having to type ';' is really nice also users don't get confused as it happens in the issue I linked above

So currently here's what you can do 
1- type ':reset' this will reset the repl then go up and modify the function
2- for single lines you can use ':pop' wich deletes last line written
3- use ':edit nano(or any editor)'

Arguably we can add a :pop_last_block but I'm not sure if its really needed

**Comment by sigmaSd** [comment]
Btw to check the state of the repl, you can use ':show'

**Comment by fdncred** [comment]
Does it check your path for executables?
```
In: :edit vscode-insiders
waiting for vscode-insiders...
The system cannot find the file specified. (os error 2)
```

**Comment by sigmaSd** [comment]
it calls `std::process::Command::new(editor)` so it should work if its in $PATH https://github.com/sigmaSd/IRust/blob/master/src/irust/parser.rs#L355

**Comment by fdncred** [comment]
`:edit notepad` works. I bet it's just looking for `.exe` in windows.

**Comment by sigmaSd** [comment]
Are you sure its on path, you can test by running code-insiders on cmd.
I just tried code and atom on linux, looks they're a bit problematic ( irust doesn't wait for them to spawn), I'm pretty sure it used to work, the problem is probably is that the initial process calls another one then exits, but anyhow this is another issue.

**Comment by fdncred** [comment]
> Are you sure its on path, you can test by running code-insiders on cmd.

yes sir. I just checked again to be sure.

**Comment by sigmaSd** [comment]
Ok I'll grab  a windows machine and debug it later.
you can test `std::process:Command::new("code-insiders").spawn().unwrap().wait()` on irust btw.
Also does it work if you use `std::process:Command::new("cmd").arg("/C").arg("code-insiders").spawn().unwrap().wait()`

**Comment by sigmaSd** [comment]
Also just a sanity check can you try `echo %PATH:;=&echo.%` and check if it contains the path to code-insiders, maybe there is a way to use executable outside of path in cmd, I'm not really sure how windows works

**Comment by fdncred** [comment]
this `std::process::Command::new("code-insiders").spawn().unwrap().wait()` does not work and give the same error message as above.
this `std::process:Command::new("cmd").arg("/C").arg("code-insiders").spawn().unwrap().wait()` does launch vscode-insiders.

**Comment by sigmaSd** [comment]
Does the second command work correctly on your case, I mean are changes reflected on irust?

**Comment by fdncred** [comment]
The second command only launches vscode, it's not hooked up to the `:edit` mechanism, if I understand what you're asking.

**Comment by sigmaSd** [comment]
for the edit command to actually work correctly:
```
In: :edit micro
waiting for micro...
Ok!
```
the ok msg needs to be printed only after the editor is closed.
Currently I noticed an issue on linux is that Command::spawn().wait() doesn't wait on forks so it means complex program like electron one wont be handled correctly, so I'm asking if its the same on windows

**Comment by fdncred** [comment]
I changed 355 to this
```rust
        std::process::Command::new("cmd").arg("/c").arg(editor)
            .arg(&*MAIN_FILE)
            .spawn()?
            .wait()?;

```
and it launches the file fine however my edits are not consumed i.e. put back in the main.rs file when i save and close.

**Comment by fdncred** [comment]
> the ok msg needs to be printed only after the editor is closed.

Nope - it says `Ok` immediately.

**Comment by fdncred** [comment]
btw - if i have `code-insiders` open, it just opens another tab with the file. I make changes to the file, save it and close the file. I don't close the editor.

**Comment by sigmaSd** [comment]
yes that's problematic, I don't see a way around closing the editor.
Does notepad work btw?
Btw I mostly use it with vi, nano, micro, maybe I should just mention its made for terminal editors, cause it looks that supporting big editors is not easy and also I doubt its really that important

**Comment by sigmaSd** [comment]
And even add a whitelist, what do you think?

**Comment by sigmaSd** [comment]
Also regarding the original issue, do you think the available ways I mentioned are enough or is there some other method you want to see added?

**Comment by sigmaSd** [comment]
Actually I don't feel strongly for the whitelist, most editors I tested works (geany, gnome-builder, gedit), I guess its really electron apps that are problematic, I think regarding this issue I can just document it

**Comment by fdncred** [comment]
> yes that's problematic, I don't see a way around closing the editor.

Perhaps what you should do is look at the last modified datetime of the file and store that in a var, then launch the editor, then when the last mod time changes, use it as if the editor closed.

> Does notepad work btw?

yes

> Btw I mostly use it with vi, nano, micro, maybe I should just mention its made for terminal editors, cause it looks that supporting big editors is not easy and also I doubt its really that important

If you're trying to be cross platform, it's not safe to assume people in windows will use vi/vim. notepad, notepad++, notepad2 are probably very safe. vscode is what a lot of people are using now cross platform so I think it'd be wise to support it.

> whitelist

white list what? editors by platform?

> original issue

I'm sitting on this at the moment. i really don't like having to have specialized code ';' in irust. I also don't like having the ability to up-arrow to change something and then it making two copies. usually, when a user hits the up-arrow in a cli app, it's to change something and since you already display the code that needs to be changed, not replacing it alltogether may be a mistake.

i like the `:edit` command, not sure what to do with the `:pop` command since it seems different than pop/stash.

**Comment by sigmaSd** [comment]
I think the issue here, that irust maybe should not accept invalid input, but this is something I really wanted to avoid since it means It will have to recompile every input instead of only the input that contains expression


>I'm sitting on this at the moment. i really don't like having to have specialized code ';' in irust. 

I actually did the same for a long time but the issue I mentioned above happened multiple times and people always got confused coming from ipython (which irust try to mimic its design to a point) that's when I decided that it must be worth it

**Comment by sigmaSd** [comment]
':pop' is handy for single line errors
1 - let x ù= "hello";
2 - :pop 
3 - up arrow, modify and enter

**Comment by sigmaSd** [comment]
> Perhaps what you should do is look at the last modified datetime of the file and store that in a var, then launch the editor, then when the last mod time changes, use it as if the editor closed.

Its tricky I can set for example a notify handler, I can see this happen asynchronously, the editor opened by the user or in fact any changes to irust/src/main.rs triggers an event that we handle by re parsing the code.

 I guess its possible, I would prefer just improving the current method though instead of the added complexity of handling async events, but its something to consider.

**Comment by sigmaSd** [comment]
@fdncred I can also workaround the issue, for example add a command ':parse_repl' (or something) so it works like this for currently broken editors
1- :edit atom
2- make the changes and save
3-  use :parse_repl which sync the repl with the modified code

The benefit is you can keep the editor open and actually make the changes whenever as long as you use the new command to sync afer

**Comment by fdncred** [comment]
i'd be willing to give that a try.

**Comment by sigmaSd** [comment]
@fdncred you can test here https://github.com/sigmaSd/IRust/tree/sync_edit unfortunately it gets pretty ugly since it reveals the internals I'll see if I can do better

**Comment by sigmaSd** [comment]
the command is :sync btw

**Comment by sigmaSd** [comment]
a bit better with https://github.com/sigmaSd/IRust/commit/5e542817c0ce4e73508ae9c59bf2b8061bda27a2

**Comment by sigmaSd** [comment]
I think I can do even better by evaluating using another file, leaving the original for cases like :edit, but the idea should be clear by now If you think this is good method I'll add it

**Comment by sigmaSd** [comment]
I pushed the changes I mentioned. Now I actually like the solution and the new command. I'll wait for your feedback before pushing this to master though.

**Comment by fdncred** [comment]
ya, i compiled and ran in the `sync_edit` branch. things look good. i can code with `code-insiders` via the `:edit` command and then when i type `:sync` it replaces my code. nice!

**Comment by sigmaSd** [comment]
cool! https://github.com/sigmaSd/IRust/pull/68
